src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,359 @@
+		if err != nil {
+			return fmt.Errorf("could not get auth token from etcdPipelineInfo: %v", err)
+		}
+		if len(resp.Kvs) != 1 {
+			return fmt.Errorf("expected to find 1 pipeline (%s), got %d: %v",
+				pipelineName, len(resp.Kvs), resp)
+		}
+		var pipelinePtr pps.EtcdPipelineInfo
+		if err := pipelinePtr.Unmarshal(resp.Kvs[0].Value); err != nil {
+			return fmt.Errorf("sidecar s3 gateway: could not unmarshal etcd pipeline info: %v", err)
+		}
+		s.pachClient.SetAuthToken(pipelinePtr.AuthToken)
+		return nil
+	}, backoff.New10sBackOff()); err != nil {
+		return fmt.Errorf("error starting sidecar s3 gateway: %v", err)
+	}
+	if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+		return nil // nothing to serve via S3 gateway
+	}
+	defer func() {
+		panic(
+			fmt.Sprintf("sidecar s3 gateway is exiting; this should never happen (err: %v)", retErr),
+		)
+	}()
+
+	return backoff.RetryNotify(func() (retErr error) {
+		return s.Serve()
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		logrus.Errorf("sidecar s3 gateway: watch closed or error monitoring jobs: %v; retrying in %v", err, d)
+		return nil
+	})
+}
+
+func (s *sidecarS3G) jobInputToBuckets(input *pps.Input) []*s3.Bucket {
+	var buckets []*s3.Bucket
+	// inputToBuckets is a recursive helper
+	pps.VisitInput(input, func(in *pps.Input) {
+		if in.Pfs != nil && in.Pfs.S3 {
+			buckets = append(buckets, &s3.Bucket{
+				Repo:   in.Pfs.Repo,
+				Commit: in.Pfs.Commit,
+				Name:   in.Pfs.Name,
+			})
+		}
+	})
+	return buckets
+}
+
+func (s *sidecarS3G) Serve() error {
+	// Watch for new jobs & initialize s3g for each new job
+	s.HandleJobs(JobHandler{
+		Begin: func(ctx context.Context, jobInfo *pps.JobInfo) error {
+			jobID := jobInfo.Job.ID
+			// Initialize new S3 gateway
+			var outputBucket *s3.Bucket
+			if s.pipelineInfo.S3Out == true {
+				outputBucket = &s3.Bucket{
+					Repo:   jobInfo.OutputCommit.Repo.Name,
+					Commit: jobInfo.OutputCommit.ID,
+					Name:   "out",
+				}
+			}
+			driver := s3.NewWorkerDriver(s.jobInputToBuckets(jobInfo.Input), outputBucket)
+
+			// server new S3 gateway & add to s.servers
+			s.serversMu.Lock()
+			defer s.serversMu.Unlock()
+			// TODO(msteffen) always serve on the same port for now (there shouldn't be
+			// more than one job in s.servers). When parallel jobs are implemented, the
+			// servers in s.servers won't actually serve anymore, and instead parent
+			// server will forward requests based on the request hostname
+			port := s.apiServer.env.S3GatewayPort
+			server, err := s3.Server(port, driver, s)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: couldn't initialize s3 gateway server: %v", err)
+			}
+			strport := strconv.FormatInt(int64(port), 10)
+			listener, err := net.Listen("tcp", ":"+strport)
+			if err != nil {
+				return fmt.Errorf("sidecar s3 gateway: could not serve on port %q: %v", port, err)
+			}
+			go func() {
+				server.Serve(listener)
+			}()
```

Comment:
1) This could be simplified with `server.ListenAndServe`.
2) This is presently ignore all errors from the HTTP server. Some errors will be for a legit state (i.e. because the server was closed), but we should probably just ignore those errors specifically.

Reply:
I meant to comment on this, but `s3g_sidecar.go` uses `ListenAndServe` now, and handles errors from it